langs/java.go

```diff
@@ -184,20 +194,71 @@ func (h *JavaLangHelper) getFDKAPIVersion() (string, error) {
 		return h.latestFdkVersion, nil
 	}
 
-	const versionURL = "https://api.bintray.com/search/packages/maven?repo=fnproject&g=com.fnproject.fn&a=fdk"
-	const versionEnv = "FN_JAVA_FDK_VERSION"
-	fetchError := fmt.Errorf("Failed to fetch latest Java FDK javaVersion from %v. Check your network settings or manually override the javaVersion by setting %s", versionURL, versionEnv)
+	const bintrayVersionURL = "https://api.bintray.com/search/packages/maven?repo=fnproject&g=com.fnproject.fn&a=fdk"
+	const mavenComVersionUrl = "https://repo1.maven.org/maven2/com/fnproject/fn/fdk/maven-metadata.xml"
+	const mavenIOVersionUrl = "https://repo1.maven.org/maven2/io/fnproject/fn/fdk/maven-metadata.xml"
 
-	type parsedResponse struct {
-		Version string `json:"latest_version"`
-	}
+	const versionEnv = "FN_JAVA_FDK_VERSION"
+	fetchError := fmt.Errorf("Failed to fetch latest Java FDK javaVersion. Check your network settings or manually override the javaVersion by setting %s", versionEnv)
 	version := os.Getenv(versionEnv)
+
 	if version != "" {
 		return version, nil
 	}
+	version, err := h.getFDKLastestFromURL(mavenComVersionUrl, mavenIOVersionUrl, bintrayVersionURL)
+	if err != nil {
+		return "", fetchError
+	}
 
-	// nishalad95: bin tray TLS certs cause verification issues on OSX, skip TLS verification
+	h.latestFdkVersion = version
+	return version, nil
+}
+
+func (h *JavaLangHelper) getFDKLastestFromURL(comURL string, ioURL string, bintrayURL string) (string, error) {
+	var buf *bytes.Buffer
+	var err error
+	err = fmt.Errorf("All urls failed to respond ")
+
+	//First search for com.fnproject.fn from Maven Central to get the latest version
+	buf, err = h.getURLResponse(comURL)
+	if err == nil {
+		version, e1 := h.parseMavenResponse(*buf)
+		if e1 == nil {
+			h.groupId = "com.fnproject.fn"
+			h.pomType = "maven"
+			return version, e1
+		}
+	}
+
+	//Second time search for io.fnproject.fn from Maven Central to get the latest version, if com.fnproject.fn fails
+	buf, err = h.getURLResponse(ioURL)
+	if err == nil {
+		version, e1 := h.parseMavenResponse(*buf)
+		if e1 == nil {
+			h.groupId = "io.fnproject.fn"
+			h.pomType = "maven"
+			return version, e1
+		}
+	}
+
+	//Third time search for com.fnproject.fn from Bintray to get the latest version, if both com.fnproject.fn and io.fnproject.fn fails
+	buf, err = h.getURLResponse(bintrayURL)
+	if err == nil {
+		version, e1 := h.parseBintrayResponse(*buf)
+		if e1 == nil {
+			h.groupId = "com.fnproject.fn"
+			h.pomType = "bintray"
+			return version, e1
+		}
+	}
+
+	//In all other case return error as latest FDK version is not identified
+	return "", err
+}
+
+func (h *JavaLangHelper) getURLResponse(url string) (*bytes.Buffer, error) {
```

Comment:
As the problem with verifying certs is only there for bintray, can we make sure that we don't use insecureSkipVerify with Maven Central? We definitely want to make sure packages downloaded from there are genuine. We could pass a flag to this function and only set it to true for the bintray path.

Reply:
Done